README.md
# Auth

> Authentication Plugin for Vue 3

- [Auth](#auth)
  - [Demo](#demo)
  - [Dependencies](#dependencies)
  - [Installation](#installation)
  - [Usage](#usage)
    - [Use as Vue Plugin](#use-as-vue-plugin)
    - [Use with Composition API](#use-with-composition-api)
      - [`login(payload: Object): Promise<AxiosResponse>`](#loginpayload-object-promiseaxiosresponse)
      - [`fetchUser(): Promise<AxiosResponse>`](#fetchuser-promiseaxiosresponse)
      - [`setToken(tokenData: string): void`](#settokentokendata-string-void)
      - [`setUser(userData: Object): void`](#setuseruserdata-object-void)
      - [`logout(): boolean`](#logout-boolean)
      - [`forceLogout(): boolean`](#forcelogout-boolean)
      - [`user: Object`](#user-object)
      - [`loggedIn: boolean`](#loggedin-boolean)
    - [Manually Creating The Auth Function](#manually-creating-the-auth-function)
  - [Default Options](#default-options)
  - [Options](#options)
    - [`endpoints`](#endpoints)
    - [`token`](#token)
    - [`user`](#user)
    - [`moduleName`](#modulename)
    - [`expiredStorage`](#expiredstorage)
    - [`redirect`](#redirect)
    - [`registerAxiosInterceptors`](#registeraxiosinterceptors)
    - [`storage`](#storage)
  - [Implementing Refresh Token](#implementing-refresh-token)
    - [Refresh Token Options](#refresh-token-options)
      - [`enabled`](#enabled)
      - [`property`](#property)
      - [`maxAge`](#maxage)
      - [`storageName`](#storagename)
      - [`name`](#name)
      - [`autoLogout`](#autologout)
  - [License](#license)

## Demo

- Checkout online demo [here](https://vue-auth-demo.vercel.app/)
- View example project [here](https://github.com/gravitano/volar-starter)

## Dependencies

- [axios](https://github.com/axios/axios): Used as the default http client.
- [Vuex](https://next.vuex.vuejs.org/): Used to store auth `state`.
- [Vou Router](https://next.vuex.vuejs.org/): Used to redirect between pages.
- [TypeScript](https://www.typescriptlang.org/): Written in TypeScript, optimize for Vue + TypeScript. But, also ship ESM and Common JS version on `dist`. 


## Installation

Install the package just like regular npm package.

```bash
npm i @gravitano/vue-auth
# OR
yarn add @gravitano/vue-auth
```

## Usage

### Use as Vue Plugin

Install the plugin to your Vue app.

```js
// main.js
import {createApp} from 'vue';
import Auth from '@gravitano/vue-auth'; // 👈 import the plugin
import store from '~/store'
import router from '~/router'
import axios from 'axios'

const app = createApp(App);

app.use(Auth); // 👈 use the plugin

// or with custom options
// 👇
app.use(Auth, {
  options, // auth options
  store, // vuex store instance
  axios, // axios instance
  router // vue router instance
}); // 👈 use the plugin

app.mount('#app');
```

After that, you can access the plugin via `$auth` global property.

```html
<template>
  <div v-if="$auth.loggedIn">Logged In as : {{ $auth.user.name }}</div>
</template>
```

If you are using composition API, you can also access the `auth` object by using `inject` method.

```js
import {injectAuth} from '@gravitano/vue-auth'

// user is Ref
const {user} = injectAuth()

// access the user
console.log(user)
```

### Use with Composition API

To use the auth in composition API, just import and use the `useAuth` function.

```ts
import {useAuth} from '@gravitano/vue-auth';

const auth = useAuth();

// OR with object destruction
const {user} = useAuth();
```

#### `login(payload: Object): Promise<AxiosResponse>`

Login the user and save token to the auth storage.

```js
auth.login({
  email: 'admin@example.com',
  password: 'admin',
});
```

#### `fetchUser(): Promise<AxiosResponse>`

Fetch user data from API.

```js
const user = await auth.fetchUser();
console.log(user);
```

#### `setToken(tokenData: string): void`

Manually Set User Token.

```js
auth.setToken(token);
```

#### `setUser(userData: Object): void`

Manually set the user data.

```js
const userData = {
  id: 1,
  name: 'Admin',
};

auth.setUser(userData);
```

#### `logout(): boolean`

Logout the current user.

```js
auth.logout();
```

#### `forceLogout(): boolean`

Force logout the current user.

```js
auth.forceLogout();
```

#### `user: Object`

Get the current user data.

```js
console.log(auth.user);
```

#### `loggedIn: boolean`

Get the current logged in state.

```js
console.log(auth.loggedIn);
```


### Manually Creating The Auth Function

First, create `auth.ts` file under your `src/plugins` folder.
```ts
import {authOptions} from '~/config';
import store, {AppRootState} from '~/store';
import {createAuth} from '@gravitano/vue-auth';
import axios from 'axios';
import router from '~/router';

export const useAuth = () => createAuth<AppRootState>(authOptions, axios, store, router);
```

Then, in your component, just import and use it as regular composition function. For example:
```vue
<script setup lang="ts">
import {useAuth} from '~/plugins/auth'

// destruct user object from `useAuth` function
const {user} = useAuth();

console.log(user); // <-- user data
</script>
```

## Default Options

This is the default options object:

```ts
import { AuthOptions } from '@gravitano/vue-auth'

export const defaultOptions: AuthOptions = {
  endpoints: {
    login: {
      url: '/auth/login',
      method: 'post',
    },
    logout: {
      url: '/auth/logout',
      method: 'delete',
    },
    user: {
      url: '/auth/me',
      method: 'get',
    },
  },
  token: {
    property: 'data.token',
    type: 'Bearer',
    storageName: 'auth.token',
    autoDecode: false,
    name: 'Authorization',
  },
  user: {
    autoFetch: true,
    property: 'data',
    storageName: 'auth.user',
  },
  moduleName: 'auth',
  expiredStorage: 'auth.expired',
  redirect: {
    home: '/',
    login: '/auth/login',
  },
  registerAxiosInterceptors: true,
  storage: {
    driver: 'secureLs', // supported: cookie, local, secureLs (secure local storage)
  },
};
```


## Options

### `endpoints`

- #### `login`
  - `url`: Login path. E.g. `/user/login`
  - `method`: HTTP Method. E.g. `GET`, `POST`, etc.
- #### `logout`
  - `url`: Logout path. E.g. `/user/logout`
  - `method`: HTTP Method. E.g. `GET`, `POST`, etc.
- #### `user`
  - `url`: Endpoint for getting user data. E.g. `/my/profile`
  - `method`: HTTP Method. E.g. `GET`, `POST`, etc.

### `token`

- #### `property`
  Token property path using dot notation.
  - Type: `string`
  - Default: `data.token`
- #### `type`
  Token type.
  - Type: `string`
  - Default: `Bearer`
- #### `storageName`
  Token storage name.
  - Type: `string`
  - Default: `auth.token`
- #### `autoDecode`
  Auto decodes token when possible. Usually used when using JWT Token.
  - Type: `boolean`
  - Default: `true`
- #### `name`
  Token header name.
  - Type: `string`
  - Default:`Authorization`

### `user`

- #### `autoFetch`
  Fetch user data automatically when user successfully logged in.
  - Type: `boolean`
  - Default: `true`
- #### `property`
  Property path of user data.
  - Type: `string`
  - Default: `data`
- #### `storageName`
  Storage name for storing user data.
  - Type: `string`
  - Default: `auth.user`

### `moduleName`
  Vuex's module name.
  - Default: `auth`
  - Type: `string`

### `expiredStorage`
  Storage name for storing token expiratin time.
  - Type: `string`
  - Default: `auth.expired`

### `redirect`
  
- #### `home`
  Homepage path.
  - Type: `string`
  - Default: `/`
- #### `login`
  Login path.
  - Type: `string`
  - Default: `/`


### `registerAxiosInterceptors`
  Register custom axios interceptors when `true`. Set the value to `false` if you want to use your own interceptors.
  - Type: `boolean`
  - Default: `true`

### `storage`

- #### `driver`
  - Type: `string`
  - Default: `secureLs`
  - Available Options: `local` | `secureLs` | `cookie`


## Implementing Refresh Token

To implement refresh token, update your auth options like so:

```diff

import { AuthOptions } from '@gravitano/vue-auth'

export const defaultOptions: AuthOptions = {
  endpoints: {
    login: {
      url: '/auth/login',
      method: 'post',
    },
    logout: {
      url: '/auth/logout',
      method: 'delete',
    },
    user: {
      url: '/auth/me',
      method: 'get',
    },
+    refresh: {
+      url: '/auth/refresh_token',
+      method: 'get',
+    },
  },
  token: {
    property: 'data.token',
    type: 'Bearer',
    storageName: 'auth.token',
    autoDecode: false,
    name: 'Authorization',
  },
  user: {
    autoFetch: true,
    property: 'data',
    storageName: 'auth.user',
  },
  moduleName: 'auth',
  expiredStorage: 'auth.expired',
  redirect: {
    home: '/',
    login: '/auth/login',
  },
  registerAxiosInterceptors: true,
  storage: {
    driver: 'secureLs', // supported: cookie, local, secureLs (secure local storage)
  },
+  refreshToken: {
+    enabled: true, // <-- make sure the value is set to true
+    property: 'data',
+    maxAge: 60 * 60 * 24 * 30, // default 30 days
+    storageName: 'auth.refresh_token',
+    name: 'refresh_token',
+    autoLogout: true,
+  },
};
```

### Refresh Token Options

#### `enabled`

Indicates the refresh token is enabled or not.

- Type: `string`
- Default: `false`


#### `property`

Path of refresh token property from the response.

- Type: `string`
- Default: `data.refresh_token`

#### `maxAge`

Maximum age of new token.

- Type: `number`
- Default: `60 * 60 * 24 * 30` (30 Days)

#### `storageName`

Storage name for storing refresh token data.

- Type: `string`
- Default: `auth.refresh_token`

#### `name`

Payload name to sent when refreshing token. 

- Type: `string`
- Default: `refresh_token`

#### `autoLogout`

When `true`, user will forced to logout and login again when refresh token failed.

- Type: `boolean`
- Default: `true`

## License

[MIT](https://opensource.org/licenses/MIT)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.2.0](https://github.com/gravitano/vue-auth/compare/v0.1.16...v0.2.0) (2022-08-23)


### Features

* vite 3 and improve lodash ([497a10c](https://github.com/gravitano/vue-auth/commit/497a10c5da94fccdc72d36b1e62cd15a750257fe))

### [0.1.16](https://github.com/gravitano/vue-auth/compare/v0.1.15...v0.1.16) (2022-03-22)


### Bug Fixes

* add `AxiosStatic` type ([efd5568](https://github.com/gravitano/vue-auth/commit/efd55689e70f76af41f92bf788f8b9e27b01e73f))

### [0.1.15](https://github.com/gravitano/vue-auth/compare/v0.1.13...v0.1.15) (2022-03-22)


### Bug Fixes

* improve typings ([df85ec1](https://github.com/gravitano/vue-auth/commit/df85ec1342e18e2581449b9ee5ee0ed9306116c0))
* pass auth options when saving state to storage ([e68b5ec](https://github.com/gravitano/vue-auth/commit/e68b5ec0aeae4221a39d8fbecb40d9682d8085ec))

### [0.1.14](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/compare/v0.0.2-alpha.1...v0.1.14) (2022-01-11)


### ⚠ BREAKING CHANGES

* merge from github
* remove unused scripts

### Features

* add `standard-version` ([04bcc06](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/04bcc06416c3b96b03d166becd8d4efc1d4e9baa))
* add axios interceptors for pinia ([8778d18](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/8778d18b7b7858aa96241bbc695b56b55a6a5166))
* add custom vue properties definition ([759bf84](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/759bf84c416177aa3fdb6b0ef0b6b643b2b2301a))
* add pinia dep ([35e9a65](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/35e9a659aa35b390a61e57780048b4990853a4c2))
* add prepublishOnly scripts ([0a497bd](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/0a497bdc1027b5dec5a5497638a846ced823d943))
* add support for refresh token ([1ce0936](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/1ce0936f879bc9bc57734dd3636c716c670358e0))
* added auth middleware ([8775609](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/8775609d2ee1bb9af0a9bb4cc33bf06b83a697b2))
* allow to send extra payload to logout method ([c875622](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/c875622bbccccc275ab6471e386090b863706a54))
* allow to set cookie options ([7ff562a](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/7ff562ab182032f1cef4412b0c8041022e83d66d))
* build and add dist ([a95dab1](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/a95dab1353155ae26dabce547f643f24961c9b12))
* configurable auth storage via auth options `storage.driver` ([cda7f3d](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/cda7f3d7cde0ba0fd6ecd17e791d56d3b95a3cea))
* export src/tyles ([ea8588c](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/ea8588c411b1510c4c0bb9138765feef4168c640))
* export utils ([723553b](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/723553beb47c7717c8fff99397bb5d95b0ef2486))
* improved error handling ([59a6d76](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/59a6d76d902249210dc9bb5737ba71f5f1acd8dc))
* improved refresh token logic ([c4b36ae](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/c4b36ae79396a36506cb83afad09a559ec753dc0))
* improved typing ([d7ea3ed](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/d7ea3ed643f2b3cbb4d3b08b3135568fd62ac4bd))
* **local storage:** always stringify the local storage value ([d7acd3e](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/d7acd3e17c0ca393c0aca0ea78164272d1a8120d))
* new `getToken` flow ([ed2a3c4](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/ed2a3c4e77102c6cfcf4f04035f122c5620288b8))
* new `getUser` function ([4953c83](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/4953c834b883b720de48fc26a349337423e74f9e))
* new composition method `loginAs` for simulating login user ([2bec70c](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/2bec70c9e738ae3447bad31d7762501fdf970047))
* new refresh token option: `autoLogout`. ([df46398](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/df463986dc0df07e08949da320ba17110e2f223a))
* new storage driver `cookie` ([ea1ee09](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/ea1ee096d747cb7a0f21e74a8eb63734a5a64d1d))
* pull origin ([6cb6c34](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/6cb6c34546af488c188bbe2232f7bcf1e917e8fd))
* remove .npmrc file ([66a9156](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/66a9156474bf30db647d5a77d5067feb3f9c88d0))
* revert back vuex composition auth ([44d9c90](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/44d9c90952e46860a3dec64a3fe3b8fcd8ad6ed5))
* ship es and other formats ([1031145](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/1031145f2ba6db86d1a909056d7f7d21ab37bd4c))
* simplify the `authModule` ([5752e0a](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/5752e0a0487f70cebb750117628cf71243b88591))
* **types:** set default types for login payload ([a65682d](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/a65682d8cebee587194897f875877df77ae43cc8))
* update `registerAxiosInterceptors` function ([6431ab9](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/6431ab989daf05834e6efed78bbdcc2f94eca80b))
* update readme ([5518a8d](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/5518a8dbd47335f6f4dff18a89d720ab18a4a167))
* wip new refresh token ([5e24ec0](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/5e24ec0fc690de88d48f403f7d75a1942e2548e2))


### Bug Fixes

* clear refresh token storage ([94b6547](https://gitlab.com/wartek-id/belajar-id/belajar.id-cms/commit/94b654745c621a1c85367a351a074cb755973856))

src/options.ts
import {AuthOptions} from './types';

export const defaultOptions: AuthOptions = {
  endpoints: {
    login: {
      url: '/auth/login',
      method: 'post',
    },
    logout: {
      url: '/auth/logout',
      method: 'delete',
    },
    user: {
      url: '/auth/me',
      method: 'get',
    },
    refresh: {
      url: '/auth/refresh',
      method: 'post',
    },
  },
  token: {
    property: 'data.token',
    type: 'Bearer',
    storageName: 'auth.token',
    autoDecode: false,
    name: 'Authorization',
  },
  user: {
    autoFetch: true,
    property: 'data',
    storageName: 'auth.user',
  },
  refreshToken: {
    enabled: false,
    property: 'data',
    maxAge: 60 * 60 * 24 * 30, // default 30 days
    storageName: 'auth.refresh_token',
    name: 'refresh_token',
    autoLogout: true,
  },
  moduleName: 'auth',
  expiredStorage: 'auth.expired',
  redirect: {
    home: '/',
    login: '/auth/login',
  },
  registerAxiosInterceptors: true,
  storage: {
    driver: 'secureLs',
  },
  baseURL: '',
};
